python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+                    s = self.A[u].head
+                    while s is not None:
+                        intvl = (
+                            max(seg.left, s.left),
+                            min(seg.right, s.right),
+                        )
+                        if intvl[1] - intvl[0] > 0:
+                            list_to_add.add(Segment(intvl[0], intvl[1], s.node))
+                        s = s.next
+
+            pn = self.calculate_ibd_segs(self.current_parent, list_to_add)
+            nodes_to_process += pn
+
+            # For parents that are also samples
+            if (self.current_parent in self.samples) and self.PARENT_SHOULD_BE_ADDED:
+                singleton_seg = SegmentList()
+                singleton_seg.add(
+                    Segment(0, self.ts.sequence_length, self.current_parent)
+                )
+                pn = self.calculate_ibd_segs(self.current_parent, singleton_seg)
+                self.PARENT_SHOULD_BE_ADDED = False
+                nodes_to_process += pn
+
+            # Move to next edge.
+            e = next(mygen, None)
+
+            # Remove any processed nodes that are no longer needed.
+            if e is None or e.parent != self.current_parent:
+                nodes_to_remove = list(set(nodes_to_process))
+                for n in nodes_to_remove:
+                    if self.current_parent in self.parent_list[n]:
+                        self.parent_list[n].remove(self.current_parent)
+                    if len(self.parent_list[n]) == 0:
+                        self.A[n] = []
+                pn = []
+
+        return self.ibd_segments
+
+    def calculate_ibd_segs(self, current_parent, list_to_add):
+        """
+        Write later.
+        """
+
+        if self.A[current_parent] is None:
+            self.A[current_parent] = list_to_add
+
+        else:
+            seg0 = self.A[current_parent].head
+            while seg0 is not None:
+                seg1 = list_to_add.head
+                while seg1 is not None:
+                    left = max(seg0.left, seg1.left)
+                    right = min(seg0.right, seg1.right)
+                    if left >= right:
+                        seg1 = seg1.next
+                        continue
+                    nodes = [seg0.node, seg1.node]
+                    nodes.sort()
+
+                    # If there are any overlapping segments, record as a new
+                    # IBD relationship.
+                    if right - left > self.min_length:
+                        self.add_ibd_segments(
+                            nodes[0],
+                            nodes[1],
+                            Segment(left, right, self.current_parent),
+                        )
+                    seg1 = seg1.next
+                seg0 = seg0.next
+
+            # Add list_to_add to A[u].
+            self.A[current_parent].add(list_to_add)
+
+        # Calculate nodes processed in this step.
```

Comment:
I'm not sure there's much point in returning this list, and it would be really tedious to do in C. All we're doing is returning a list consisting of the nodes in the input segment list, right?

Reply:
So, this list does serve a purpose here -- it's how we figure out which nodes' parents have been completely processed at a given step of the algorithm, and so which segments can be cleared from memory.  I'll think about whether there are better ways of doing this.